File: Python/0380_InsertDeleteGetRandomO(1).py
import random

# Solution 1 - Hash Map and List
class RandomizedSet:
    def __init__(self):
        """
        Initialize your data structure here.
        """
        self.map = {}
        self.arr = []

    def insert(self, val: int) -> bool:
        """
        Inserts a value to the set. Returns true if the set did not already contain the specified element.
        """
        if val in self.map:
            return False
        self.map[val] = len(self.arr)
        self.arr.append(val)
        return True

    def remove(self, val: int) -> bool:
        """
        Removes a value from the set. Returns true if the set contained the specified element.
        """
        if val not in self.map:
            return False
        # idx is the to-delete value index in arr
        # last is the last value in the arr
        idx, last = self.map[val], self.arr[-1]
        # set the value at to-delete index to the last value in arr
        # set the idx of the last value in arr in map to the new idx
        # note the new idx is the idx of the to-delete value
        self.arr[idx], self.map[last] = last, idx
        self.arr.pop()
        self.map.pop(val, 0)
        return True

    def getRandom(self) -> int:
        """
        Get a random element from the set.
        """
        return self.arr[random.randint(0, len(self.arr)-1)]


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()

File: C++/0380_InsertDeleteGetRandomO(1).cpp
class RandomizedSet
{
  private:
    vector<int> arr;
    unordered_map<int, int> m_;

  public:
    /** Initialize your data structure here. */
    RandomizedSet() {}

    /** Inserts a value to the set. Returns true if the set did not already
   * contain the specified element. */
    bool insert(int val)
    {
        if (m_.count(val))
            return false;
        m_[val] = arr.size();
        arr.push_back(val);
        return true;
    }

    /** Removes a value from the set. Returns true if the set contained the
   * specified element. */
    bool remove(int val)
    {
        if (!m_.count(val))
            return false;
        int idx = m_[val];
        m_[arr.back()] = idx;
        m_.erase(val);
        std::swap(arr[idx], arr.back());
        arr.pop_back();
        return true;
    }

    /** Get a random element from the set. */
    int getRandom() { return arr[rand() % arr.size()]; }
};

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet* obj = new RandomizedSet();
 * bool param_1 = obj->insert(val);
 * bool param_2 = obj->remove(val);
 * int param_3 = obj->getRandom();
 */
